### PR TITLE
feat: fallbacks and guards

### PR DESCRIPTION
**Pull request by xBA5ED**
# Description

*What does this PR: do, how, why?*

## Limitations & risks

*Are there any trade-off or new vulnarbility surface based on theses changes?*

# Check-list
- [ ] Tests are covering the new feature
- [ ] Code is [natspec'd](https://docs.soliditylang.org/en/v0.8.17/natspec-format.html)
- [ ] Code is [linted and formatted](https://docs.soliditylang.org/en/v0.8.17/style-guide.html)
- [ ] I have run the test locally (and they pass)
- [ ] I have rebased to the latest main commit (and tests still pass)

# Interactions
These changes will impact the following contracts:
- Directly:

- Indirectly:

### CODE REVIEW

File: src/structs/JBRemoteToken.sol [hunk 1]
```diff
@@ -2,11 +2,15 @@
 pragma solidity ^0.8.0;
 
 /// @notice A struct that represents a token on the remote chain.
+/// @dev Invarient: If the `emergencyHatch` is true then the `enabled` is always false.
+/// @custom:member enabled Whether the token is enable.
```
Comment (mejango) [inline src/structs/JBRemoteToken.sol]: nit: "enabled"

File: src/structs/JBRemoteToken.sol [hunk 2]
```diff
@@ -2,11 +2,15 @@
...
+/// @custom:member emergencyHatchOpened Whether the emergency hatch is opened.
```
Comment (mejango) [inline src/structs/JBRemoteToken.sol:7]: maybe "forceDisable"

File: src/interfaces/IJBSuckerRegistry.sol [hunk 3]
```diff
@@ -18,6 +20,8 @@ interface IJBSuckerRegistry {
     function suckerDeployerIsAllowed(address deployer) external view returns (bool);
 
     function allowSuckerDeployer(address deployer) external;
+    function removeSuckerDeployer(address deployer) external;
```
Comment (mejango) [inline src/interfaces/IJBSuckerRegistry.sol]: nit: alphebetize function definitions

File: src/interfaces/IJBSucker.sol [hunk 4]
```diff
@@ -53,6 +55,7 @@ interface IJBSucker is IERC165 {
 
     function peerChainId() external view returns (uint256 chainId);
     function isMapped(address token) external view returns (bool);
+    function deprecated() external view returns (JBSuckerDeprecationState);
```
Comment (mejango) [inline src/interfaces/IJBSucker.sol]: nit: alphebetize view function defs

File: src/enums/JBSuckerDeprecationState.sol [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+/// @notice Options for the deprecation state of a `JBSucker`.
+/// @custom:element NOT_DEPRECATED The `JBSucker` is not deprecated.
```
Comment (mejango) [inline src/enums/JBSuckerDeprecationState.sol]: we use custom:member throughout

File: src/JBSuckerRegistry.sol [hunk 6]
```diff
@@ -219,6 +231,27 @@ contract JBSuckerRegistry is Ownable, JBPermissioned, IJBSuckerRegistry {
...
     }
 
+    /// @notice Lets anyone remove a deprecated sucker from a project.
+    /// @param projectId The ID of the project to remove the sucker from.
+    /// @param sucker The address of the deprecated sucker to remove.
+    function removeDeprecatedSucker(uint256 projectId, IJBSucker sucker) public {
+        // Sanity check, make sure that the sucker does actually belong to the project.
+        (bool belongsToProject, uint256 val) = _suckersOf[projectId].tryGet(address(sucker));
+        if (!belongsToProject || val != _SUCKER_EXISTS) {
+            revert JBSuckerRegistry_SuckerDoesNotBelongToProject();
```
Comment (mejango) [inline src/JBSuckerRegistry.sol]: maybe add the address as an error arg?

File: src/JBSuckerRegistry.sol [hunk 7]
```diff
@@ -219,6 +231,27 @@ contract JBSuckerRegistry is Ownable, JBPermissioned, IJBSuckerRegistry {
...
+        }
+
+        // Check if the sucker is deprecated.
+        JBSuckerDeprecationState state = sucker.deprecated();
```
Comment (mejango) [inline src/JBSuckerRegistry.sol]: maybe call this fn "state()" instead of "deprecated()"?

File: src/enums/JBSuckerDeprecationState.sol [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+/// @custom:element DEPRECATION_PENDING The `JBSucker` has a deprecation set, but it is still fully functional.
+/// @custom:element SENDING_DISABLED The `JBSucker` is deprecated and sending to the pair sucker is disabled.
+/// @custom:element DEPRECATED The `JBSucker` is deprecated, but it continues to let users claim their funds.
+enum JBSuckerDeprecationState {
```
Comment (mejango) [inline src/enums/JBSuckerDeprecationState.sol]: maybe just "JBSuckerState"

File: src/enums/JBSuckerDeprecationState.sol [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+    NOT_DEPRECATED,
```
Comment (mejango) [inline src/enums/JBSuckerDeprecationState.sol]: "ENABLED"

File: src/JBSucker.sol [hunk 10]
```diff
@@ -113,6 +117,9 @@ abstract contract JBSucker is JBPermissioned, ERC165, IJBSucker {
     // -------------------- internal stored properties ------------------- //
     //*********************************************************************//
 
+    /// @notice The timestamp after which the sucker is entirely deprecated.
+    uint40 internal deprecatedAfter;
```
Comment (mejango) [inline src/JBSucker.sol]: can be uint256 for consistency

File: src/JBSucker.sol [hunk 11]
```diff
@@ -354,6 +379,11 @@ abstract contract JBSucker is JBPermissioned, ERC165, IJBSucker {
         address token = map.localToken;
         JBRemoteToken memory currentMapping = _remoteTokenFor[token];
 
+        // Once the emergency hatch for a token is enabled it can't be disabled.
+        if (currentMapping.emergencyHatch) {
+            revert JBSucker_TokenHasInvalidEmergencyHatchState(token);
```
Comment (mejango) [inline src/JBSucker.sol:400]: TokenIsForcedDisabled

File: src/JBSucker.sol [hunk 12]
```diff
@@ -407,6 +446,28 @@ abstract contract JBSucker is JBPermissioned, ERC165, IJBSucker {
...
+        // The caller must be the project owner or have the `QUEUE_RULESETS` permission from them.
+        // slither-disable-next-line calls-loop
+        _requirePermissionFrom({
+            account: DIRECTORY.PROJECTS().ownerOf(PROJECT_ID),
+            projectId: PROJECT_ID,
+            permissionId: JBPermissionIds.MAP_SUCKER_TOKEN
+        });
+
+        // Enable the emergency hatch for each token.
+        for (uint256 _i; _i < tokens.length; _i++) {
```
Comment (mejango) [inline src/JBSucker.sol]: no underline for i

File: src/JBSucker.sol [hunk 13]
```diff
@@ -529,6 +593,62 @@ abstract contract JBSucker is JBPermissioned, ERC165, IJBSucker {
...
+    /// @notice Set or remove the time after which this sucker will be deprecated, once deprecated the sucker will no
+    /// longer be functional and it will let all users exit.
+    /// @param timestamp The time after which the sucker will be deprecated. Or `0` to remove the upcoming deprecation.
+    function setDeprecation(uint40 timestamp) external {
+        // As long as the sucker has not reached the final deprecation state its deprecation time can be
+        // extended/shortened.
+        if (deprecated() == JBSuckerDeprecationState.DEPRECATED) revert JBSucker_Deprecated();
+
+        // slither-disable-next-line calls-loop
+        // TODO: Change permissionId?
```
Comment (mejango) [inline src/JBSucker.sol]: lets add permission IDs for this

File: src/JBSucker.sol [hunk 14]
```diff
@@ -529,6 +593,62 @@ abstract contract JBSucker is JBPermissioned, ERC165, IJBSucker {
...
+        _requirePermissionFrom({
+            account: DIRECTORY.PROJECTS().ownerOf(PROJECT_ID),
+            projectId: PROJECT_ID,
+            permissionId: JBPermissionIds.MAP_SUCKER_TOKEN
+        });
+
+        // This is the earliest time for when the sucker can be considered deprecated.
+        // There is a mandatory delay to allow for remaining messages to be received.
+        // This should be called on both sides of the suckers, preferably with a matching timestamp.
+        uint40 nextEarliestDeprecationTime = uint40(block.timestamp) + _maxMessagingDelay();
```
Comment (mejango) [inline src/JBSucker.sol]: do we need uint4o? can we just do default 256?

File: src/JBSucker.sol [hunk 15]
```diff
@@ -650,7 +799,67 @@ abstract contract JBSucker is JBPermissioned, ERC165, IJBSucker {
...
+        uint256 amount,
+        JBRemoteToken memory remoteToken,
+        JBMessageRoot memory message
+    )
+        internal
+        virtual;
+
+    /// @notice What is the maximum time it takes for a message to be received on the other side.
+    /// @dev Be sure to keep in mind if a message fails having to retry and the time it takes to retry.
+    function _maxMessagingDelay() internal view virtual returns (uint40) {
```
Comment (mejango) [inline src/JBSucker.sol]: can this be a constant?

Comment (xBA5ED) [reply src/JBSucker.sol]: The `_maxMessagingDelay` depends on the bridge that a sucker uses, so each sucker should be able to specify how long a message might take. 14 days is just a default value thats likely safe for all suckers.

**Review by mejango — APPROVED** [review]

**Pull request merged**